### ISSUE TITLE
VB6 App object not supported

### ISSUE THREAD

**Issue by dmrvb** [WaynePhillipsEA/twinbasic#143]
**Describe the bug**
The VB6 APP object is not available, 

**To Reproduce**
APP.path etc. gives an unrecognized symbol warning:

maybe this is because APP is related to the file system in some way???

**Desktop (please complete the following information):**
 - OS: Windows 10
 - twinBASIC compiler version 0.9.2720

**Comment by WaynePhillipsEA** [comment]
Yes, this will be implemented soon.  Thanks for reporting.

**Comment by Kr00l** [comment]
Would it be possible to have the App object already partially ? Maybe it's difficult for the moment to provide 100% for the App object in one shot. But maybe it would be quite quick (?) to provide at least `App.hInstance` and `App.ThreadID`.
This would allow more testing possibilities for some API functions/callbacks etc.

**Comment by WaynePhillipsEA** [comment]
Yes, I think it's a good idea to start offering the simple App object methods - I'll get onto it.

**Comment by WaynePhillipsEA** [comment]
@Kr00l v0.10.4563 now includes an early implementation of the App object, offering just the hInstance and ThreadID members as requested.  Currently the App object is a predeclared class, but will later move to be a member of the Global predeclared class instead, when that exists, to match VB6.

**Comment by Kr00l** [comment]
The `App.hInstance` returns currently `As Long`.
But I think it is safe to use `As LongPtr`? ( like it's done for VarPtr etc.)

**Comment by WaynePhillipsEA** [comment]
They are internally defined as ULONG_PTR / LONG_PTR, but the type-library builder for the internal libraries only outputs Long / LongLong at the moment.  I'll get that corrected.  BTW same behaviour for VarPtr/ObjPtr etc.

**Comment by WaynePhillipsEA** [comment]
App.Path and App.EXEName are now supported, as of v0.13.34

**Comment by Kr00l** [comment]
> App.Path and App.EXEName are now supported, as of v0.13.34

Also long path aware? :-)

**Comment by Kr00l** [comment]
I saw that App.IsInIDE property is added. Why not enhancing the VB6 `App.StartupMode` by adding an enum `vbSModeInIDE`.
This would save an additional property.

**Comment by WaynePhillipsEA** [comment]
> I saw that App.IsInIDE property is added. Why not enhancing the VB6 `App.StartupMode` by adding an enum `vbSModeInIDE`. This would save an additional property.

Good idea, but that could break backwards compatibility where the StartupMode is expected to only be one of the original two enum values.

**Comment by Kr00l** [comment]
> > I saw that App.IsInIDE property is added. Why not enhancing the VB6 `App.StartupMode` by adding an enum `vbSModeInIDE`. This would save an additional property.
> 
> Good idea, but that could break backwards compatibility where the StartupMode is expected to only be one of the original two enum values.

Ouch. Yes. :-)

**Comment by WaynePhillipsEA** [comment]
As mentioned in #608 by @jpbro, the following members are not yet supported:

- App.Comments
- App.CompanyName
- App.FileDescription
- App.HelpFile
- App.LegalCopyright
- App.LegalTrademarks
- App.LogEvent
- App.LogMode
- App.LogPath
- App.Major
- App.Minor
- App.NonModalAllowed
- App.OleRequestPendingMsgText
- App.OleRequestPendingMsgTitle
- App.OleRequestPendingTimeout
- App.OleServerBusyMsgText
- App.OleServerBusyMsgTitle
- App.OleServerBusyRaiseError
- App.OleServerBusyTimeout
- App.Revision
- App.PreviousInstance
- App.ProductName
- App.RetainedProject
- App.Revision
- App.StartLogging
- App.StartMode
- App.TaskVisible
- App.Title
- App.UnattendedApp

**Comment by Kr00l** [comment]
When `App.Major`, `App.Minor` and `App.Revision` will be supported I would vote to also include an `App.Build` which is missing in VB6.

The properties are constructed from VS_FIXEDFILEINFO like this:

HiWord(dwFileVersionMS) = App.Major
LoWord(dwFileVersionMS) = App.Minor
HiWord(dwFileVersionLS) = **App.Build**
LoWord(dwFileVersionLS) = App.Revision

**Comment by bclothier** [comment]
I'm unfamiliar with the behavior of `App.Revision` -- does it auto-increment?

I think it'd be nice if either the `Build` or `Revision` allows for auto-increment as to help track the builds.

**Comment by Kr00l** [comment]
> I'm unfamiliar with the behavior of `App.Revision` -- does it auto-increment?
> 
> I think it'd be nice if either the `Build` or `Revision` allows for auto-increment as to help track the builds.

VB6 allows to auto increment.
![image](https://user-images.githubusercontent.com/33756666/147674314-46e62ddc-8b47-492c-b25a-eef0d92bc7a1.png)

The App.Revision increments by 1 on each compilation.

Of course tB could be more flexible here in regards to `App.Build` and `App.Revision`.

**Comment by Kr00l** [comment]
According to MS [version class](https://docs.microsoft.com/en-us/dotnet/api/system.version?redirectedfrom=MSDN&view=net-6.0)

`Build: A difference in build number represents a recompilation of the same source. Different build numbers might be used when the processor, platform, or compiler changes.`

So, tB shall auto-increment also only the Revision. The Build could come in place for either x64/x86 mode or cross-platform?

**Comment by Kr00l** [comment]
VB6 always put build number **0** in the outputs. e.g. 1.0.**0**.2 for above screenshot.
So, it's not very clear. Shall tB just allow to define the build number by the developer or should there be some semantics there by tB ?

**Comment by mansellan** [comment]
Wanted to mention [Semantic Versioning](https://semver.org/), even though I don't see how it could be accommodated in this structure.

**Comment by WaynePhillipsEA** [comment]
As of v0.13.76, App.Major, App.Minor and App.Revision (plus App.Build) are now supported.

**Comment by Kr00l** [comment]
Is the "ThunderMain" window yet in tB?
When so, I would propose to expose it in a read-only property in the App object.
Naming suggestion would be just `App.hWnd`

**Comment by OldestCoderInTown** [comment]
Do we have any estimate for App.Title becoming a thing we can use?

I have a neat wee bit of code to ensure taht only ione instance of my programs run, and it depends on App.Title to identify the program

**Comment by vbBaka** [comment]
so, APP.path is not working properly in VB6. and u can not fix it because of compatibility.
it can not read certain character and it can't handle longpath properly. example:
folderø4_!   555þõｽｾ
this will not return something readable. so thats why I use a longpath API and convert to shortpath using API
so I can use VB6 filesystem.

will APP.path have parameters?
I see in TB, u have APP.path()
so it can return 8.3 foldername? like APP.path(1)

also APP.PrevInstance should be priority, as that is used to stop you from running the same app multiple times.

**Comment by WaynePhillipsEA** [comment]
`App.PrevInstance` is already supported.

**Comment by bclothier** [comment]
JFYI, converted [the list of missing members into a task list](https://github.com/twinbasic/twinbasic/issues/143#issuecomment-1002434019) to aid in tracking since a number of them got implemented since the post.

**Comment by mburns08109** [comment]
@Ben, you need to go back and add App.Build to that tasklist (checked already per Wayne) - if only for completeness

**Comment by bclothier** [comment]
Technically, the `App.Build` never existed in the original VB6 `App` - the list is meant to enumerate the missing members from the original VB6 `App` object. I believe tB's `App` has a number of other members that were added and don't exist in VB6 besides `Build`.

**Comment by WaynePhillipsEA** [comment]
App.Title is now implemented, as of BETA 132.

**Comment by dmrvb** [comment]
App.HelpFile not yet implemented

**Comment by dmrvb** [comment]
App.Helpfile ,App.Startmode and App.PrevInstance are now the only non-GUI items preventing my big apps from compiling, exciting stuff.

**Comment by Kr00l** [comment]
> App.Title is now implemented, as of BETA 132.

The `App.Title` is read-only in tB. In VB6 it can be changed at run-time. Is that intended for the time being or a bug ?

And if changed at run-time.. the `AppActivate` method will respect the new title.

**Comment by WaynePhillipsEA** [comment]
> > App.Title is now implemented, as of BETA 132.
> 
> The `App.Title` is read-only in tB. In VB6 it can be changed at run-time. Is that intended for the time being or a bug ?
> 
> And if changed at run-time.. the `AppActivate` method will respect the new title.

Yeah, I think that was just missed.   It's now on the list :)  Thanks

**Comment by WaynePhillipsEA** [comment]
App.Title can now be set at runtime, as of BETA 174.  Thanks.

**Comment by WaynePhillipsEA** [comment]
App.HelpFile is now available, as of BETA 190

**Comment by marcoslm** [comment]
It seems that support for the 'App' object has 'disappeared' in a console-type project. I have an application that compiled without issues in previous BETA versions (2xx), but now I'm encountering the problem "Unrecognized Symbol 'App'".
I'm not sure from which version this started happening.

![imagen](https://github.com/twinbasic/twinbasic/assets/2956806/37708a69-af61-4acb-b4c3-1549f16dc66d)

**Comment by sokinkeso** [comment]
> It seems that support for the 'App' object has 'disappeared' in a console-type project. I have an application that compiled without issues in previous BETA versions (2xx), but now I'm encountering the problem "Unrecognized Symbol 'App'". I'm not sure from which version this started happening.
> 
> ![imagen](https://user-images.githubusercontent.com/2956806/244782221-37708a69-af61-4acb-b4c3-1549f16dc66d.png)

The App object exists in WinNativeForms package. You can add the reference to your project from the settings

**Comment by marcoslm** [comment]
> The App object exists in WinNativeForms package. You can add the reference to your project from the settings

I appreciate your help in clarifying the issue, sokineso. I'll make sure to add the WinNativeForms package reference to my project settings. Thanks!

**Comment by fafalone** [comment]
We really need [Unimplemented] stubs instead of all these missing methods being hard errors. I hope this helps... I've taken the _App interface definition, and put in [Unimplemented] entries for all the missing members, and arranged everything in the correct vtable order.

```vba
Public Interface _App Extends stdole.IUnknown      ' FIXME needs to suppot IDispatch
    [RedirectToStaticImplementation("InternalStuff.GetAppPath")]
    Property Get Path() As String
    [RedirectToStaticImplementation("InternalStuff.GetAppEXEName")]
    Property Get EXEName() As String
    [RedirectToStaticImplementation("InternalStuff.GetAppTitle")]
    Property Get Title() As String
    [RedirectToStaticImplementation("InternalStuff.LetAppTitle")]
    Property Let Title(ByVal Value As String)
    [RedirectToStaticImplementation("InternalStuff.GetAppPrevInstance")]
    Property Get PrevInstance() As Boolean
    [Unimplemented]
    Property Get StartMode() As Integer
    [Unimplemented]
    Property Get TaskVisible() As Boolean
    [Unimplemented]
    Property Let TaskVisible(ByVal value As Boolean)
    [Unimplemented]
    Property Get OleServerBusyTimeout() As Long
    [Unimplemented]
    Property Let OleServerBusyTimeout(ByVal value As Long)
    [Unimplemented]
    Property Get OleServerBusyMsgTitle() As String
    [Unimplemented]
    Property Let OleServerBusyMsgTitle(ByVal value As String)
    [Unimplemented]
    Property Get OleServerBusyMsgText() As String
    [Unimplemented]
    Property Let OleServerBusyMsgText(ByVal value As String)
    [Unimplemented]
    Property Get OleServerBusyRaiseError() As Boolean
    [Unimplemented]
    Property Let OleServerBusyRaiseError(ByVal value As Boolean)
    [Unimplemented]
    Property Get OleRequestPendingTimeout() As Long
    [Unimplemented]
    Property Let OleRequestPendingTimeout(ByVal value As Long)
    [Unimplemented]
    Property Get OleRequestPendingMsgTitle() As String
    [Unimplemented]
    Property Let OleRequestPendingMsgTitle(ByVal value As String)
    [Unimplemented]
    Property Get OleRequestPendingMsgText() As String
    [Unimplemented]
    Property Let OleRequestPendingMsgText(ByVal value As String)
    [RedirectToStaticImplementation("InternalStuff.GetAppMajor")]
    Property Get Major() As Integer
    [RedirectToStaticImplementation("InternalStuff.GetAppMinor")]
    Property Get Minor() As Integer
    [RedirectToStaticImplementation("InternalStuff.GetAppRevision")]
    Property Get Revision() As Integer
    [RedirectToStaticImplementation("InternalStuff.GetAppComments")]
    Property Get Comments() As String
    [RedirectToStaticImplementation("InternalStuff.GetAppCompanyName")]
    Property Get CompanyName() As String
    [RedirectToStaticImplementation("InternalStuff.GetAppFileDescription")]
    Property Get FileDescription() As String
    [RedirectToStaticImplementation("InternalStuff.GetAppLegalCopyright")]
    Property Get LegalCopyright() As String
    [RedirectToStaticImplementation("InternalStuff.GetAppLegalTrademarks")]
    Property Get LegalTrademarks() As String
    [RedirectToStaticImplementation("InternalStuff.GetAppProductName")]
    Property Get ProductName() As String
    [RedirectToStaticImplementation("InternalStuff.GetApphInstance")]
    Property Get hInstance() As LongPtr
    [Unimplemented]
    Property Get NonModalAllowed() As Boolean
    [Unimplemented]
    Property Get LogPath() As String
    [RedirectToStaticImplementation("InternalStuff.GetAppLogMode")]
    Property Get LogMode() As LogModeConstants      ' FIXME Currently only supports vbLogOff and vbLogAuto for IDE detection purposes.
    [Unimplemented]
    Property Get UnattendedApp() As Boolean
    [RedirectToStaticImplementation("InternalStuff.GetAppThreadID")]
    Property Get ThreadID() As Long
    [RedirectToStaticImplementation("InternalStuff.GetAppHelpFile")]
    Property Get HelpFile() As String
    [RedirectToStaticImplementation("InternalStuff.LetAppHelpFile")]
    Property Let HelpFile(ByVal Value As String)
    [Unimplemented]
    Property Get RetainedProject() As Boolean
    [Unimplemented]
    Sub StartLogging(ByVal LogTarget As String, ByVal LogModes As Long)
    [Unimplemented]
    Sub LogEvent(ByVal LogBuffer As String, ByVal EventType As Variant)
    [RedirectToStaticImplementation("InternalStuff.GetAppIsInIDE")]
    Property Get IsInIDE() As Boolean
    [RedirectToStaticImplementation("InternalStuff.GetAppBuild")]
    Property Get Build() As Integer
    [RedirectToStaticImplementation("InternalStuff.GetAppModulePath")]
    Property Get ModulePath() As String
    [RedirectToStaticImplementation("InternalStuff.GetAppLastBuildPath")]
    Property Get LastBuildPath() As String
End Interface
```

**Comment by wqweto** [comment]
Btw, missing the same stubs.

Marking these w/ `[Unimplemented]` also brings the warm fuzzy feeling that these **will** eventually get implemented for v1 which is important :-))

**Comment by SuperDre** [comment]
> When `App.Major`, `App.Minor` and `App.Revision` will be supported I would vote to also include an `App.Build` which is missing in VB6.

Actually, what is missing from the version is **Release** not **Build**. VB6 always had **Revision** instead of **Build**, but was missing **Release**.
A version structure is made up by Major.Minor.Release.Build/Revision.
So adding a new property **Release** would be the way to go.
BUT maybe also add a new property **Build** which would just be an alternative for **Revision** and both would point to the same underlaying internal variable, so keeping **Revision** for backward compatibility (but then again, just a one time search and replace isn't a big deal either).

**Comment by fafalone** [comment]
`App.Build` and associated field in properties has already been added... and where do you see that Windows describes that as release instead? Revision is actually the one I don't see anywhere in the structs or documentation for version but that's decided for us by compatibility concerns.

**Comment by WaynePhillipsEA** [comment]
Missing App members are now marked as `[Unimplemented]` in BETA 767, thanks @fafalone
